tensorcircuit/templates/lattice.py

```diff
@@ -0,0 +1,1539 @@
+        self._identifiers = list(identifiers)
+        self._coordinates = [np.array(c) for c in coordinates]
+        self._indices = list(range(len(identifiers)))
+        self._ident_to_idx = {ident: idx for idx, ident in enumerate(identifiers)}
+
+        # Validate coordinate dimensions
+        for i, coord in enumerate(self._coordinates):
+            if coord.shape != (dimensionality,):
+                raise ValueError(
+                    f"Coordinate at index {i} has shape {coord.shape}, "
+                    f"expected ({dimensionality},)"
+                )
+
+        logger.info(f"CustomizeLattice with {self.num_sites} sites created.")
+
+        if precompute_neighbors is not None and precompute_neighbors > 0:
+            self._build_neighbors(max_k=precompute_neighbors)
+
+    def _build_lattice(self, *args: Any, **kwargs: Any) -> None:
+        """For CustomizeLattice, lattice data is built during __init__."""
+        pass
+
+    def _build_neighbors(self, max_k: int = 1, **kwargs: Any) -> None:
+        """Calculates neighbors using a KDTree for efficiency.
+
+        This method avoids storing the full N x N distance matrix. It first
+        uses `scipy.spatial.distance.pdist` to efficiently calculate all
+        unique pairwise distances to identify the neighbor shells. Then, it
+        builds a `scipy.spatial.KDTree` for fast radius queries. For each
+        site, it queries for points within each distance shell and subtracts
+        the points from inner shells to find the exact neighbors for a given k.
+        This approach is more memory-efficient and faster for large N.
+
+        If Scipy is not installed, it gracefully falls back to the slower,
+        distance-matrix-based method from the base class.
+
+        :param max_k: The maximum number of neighbor shells to
+            calculate. Defaults to 1.
+        :type max_k: int, optional
+        :param tol: The numerical tolerance for distance
+            comparisons. Defaults to 1e-6.
+        :type tol: float, optional
+        """
+        tol = kwargs.get("tol", 1e-6)
+        logger.info(f"Building neighbors for CustomizeLattice up to k={max_k}...")
+        if self.num_sites < 2:
+            return
+
+        if not SCIPY_AVAILABLE:
+            logger.warning(
+                "Scipy is not installed. Falling back to the slower, "
+                "distance-matrix-based neighbor finding method. "
+                "For better performance, please install it using 'pip install scipy'."
+            )
+            # Fallback to the original, robust-but-slower method
+            return self._build_neighbors_by_distance_matrix(max_k=max_k, tol=tol)
+
+        all_coords = np.array(self._coordinates)
+
+        # 1. Use pdist for memory-efficient calculation of pairwise distances
+        #    to robustly identify the distance shells.
+        all_distances_sq = pdist(all_coords, metric="sqeuclidean")
```

Comment:
`self._distance_matrix` assignment for CustomizeLattice KDTree engine case

Reply:
again `distance_matrix` should be assembled from `all_distances_sq`, otherwise, the all-to-all distance is computed twice?

Reply:
distance matrix should also be assigned here, so that after neighbor building, you dont need to recompute `pdist` by calling `_compute_distance_matrix`. The logic, lattice class without neighbor building, distance matrix can be accessed by `_compute_distance_matrix`, lattice class after neighbor building can directly return the matrix without recompute via `pdist` in `_compute_distance_matrix`